core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -151,20 +164,28 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
         .run(
             manifest -> {
               try (ManifestReader<?> reader = ManifestFiles.open(manifest, io)) {
+                List<String> pathsToDelete = Lists.newArrayList();
                 for (ManifestEntry<?> entry : reader.entries()) {
                   // intern the file path because the weak key map uses identity (==) instead of
                   // equals
                   String path = entry.file().path().toString().intern();
                   Boolean alreadyDeleted = deletedFiles.putIfAbsent(path, true);
                   if (alreadyDeleted == null || !alreadyDeleted) {
                     try {
-                      io.deleteFile(path);
+                      if (io instanceof SupportsBulkOperations) {
+                        pathsToDelete.add(path);
+                      } else {
+                        io.deleteFile(path);
+                      }
                     } catch (RuntimeException e) {
                       // this may happen if the map of deleted files gets cleaned up by gc
                       LOG.warn("Delete failed for data file: {}", path, e);
                     }
                   }
                 }
+                if (io instanceof SupportsBulkOperations) {
```

Comment:
this should be a part of try catch?

Reply:
My rationale was for ensuring best effort we already have an onFailure + suppressFailureWhenFinished hook at the task level in case a bulk deletion failure gets throw. However, after a 2nd pass, the error message is a bit confusing in the context of batch deletion failure "Failed to get deleted files: this may cause orphaned data files". Since at the task level we are parallelizing over reading the manifests, it's only expecting reading the manifest to be a potential failure.

So to avoid confusing messaging, I think it makes sense now to explicitly catch BulkDeletionFailure, log it and move on.